File: arkhe_multilang/ucd/ucd.go
// ucd.go
package main

import (
	"fmt"
	"math"
)

type UCD struct {
	data [][]float64
	C    float64
	F    float64
}

func NewUCD(data [][]float64) *UCD {
	return &UCD{data: data}
}

func (u *UCD) pearson(x, y []float64) float64 {
	n := len(x)
	sumX, sumY := 0.0, 0.0
	for i := 0; i < n; i++ {
		sumX += x[i]
		sumY += y[i]
	}
	meanX := sumX / float64(n)
	meanY := sumY / float64(n)
	var num, denX, denY float64
	for i := 0; i < n; i++ {
		dx := x[i] - meanX
		dy := y[i] - meanY
		num += dx * dy
		denX += dx * dx
		denY += dy * dy
	}
	if denX == 0 || denY == 0 {
		return 1.0
	}
	return num / math.Sqrt(denX*denY)
}

func (u *UCD) Analyze() (C, F float64, conservation bool, topology, scaling string, optimization float64) {
	n := len(u.data)
	if n > 1 {
		var sumCorr float64
		for i := 0; i < n; i++ {
			for j := 0; j < n; j++ {
				corr := u.pearson(u.data[i], u.data[j])
				sumCorr += math.Abs(corr)
			}
		}
		u.C = sumCorr / float64(n*n)
	} else {
		u.C = 0.5
	}
	u.F = 1.0 - u.C
	conservation = math.Abs(u.C+u.F-1.0) < 1e-10
	if u.C > 0.8 {
		topology = "toroidal"
	} else {
		topology = "other"
	}
	if u.C > 0.7 {
		scaling = "self-similar"
	} else {
		scaling = "linear"
	}
	optimization = u.F * 0.5
	return u.C, u.F, conservation, topology, scaling, optimization
}

func main() {
	data := [][]float64{
		{1, 2, 3, 4},
		{2, 3, 4, 5},
		{5, 6, 7, 8},
	}
	ucd := NewUCD(data)
	c, f, cons, topo, scale, opt := ucd.Analyze()
	fmt.Printf("C: %.4f, F: %.4f, conservation: %v, topology: %s, scaling: %s, opt: %.4f\n",
		c, f, cons, topo, scale, opt)
}

File: arkhe_multilang/ucd/ucd.rs
// ucd.rs
use std::f64;

struct UCD {
    data: Vec<Vec<f64>>,
    c: f64,
    f: f64,
}

impl UCD {
    fn new(data: Vec<Vec<f64>>) -> Self {
        UCD { data, c: 0.0, f: 0.0 }
    }

    fn pearson(&self, x: &[f64], y: &[f64]) -> f64 {
        let n = x.len();
        let mean_x = x.iter().sum::<f64>() / n as f64;
        let mean_y = y.iter().sum::<f64>() / n as f64;
        let mut num = 0.0;
        let mut den_x = 0.0;
        let mut den_y = 0.0;
        for i in 0..n {
            let dx = x[i] - mean_x;
            let dy = y[i] - mean_y;
            num += dx * dy;
            den_x += dx * dx;
            den_y += dy * dy;
        }
        if den_x == 0.0 || den_y == 0.0 { 1.0 } else { num / (den_x * den_y).sqrt() }
    }

    fn analyze(&mut self) -> (f64, f64, bool, String, String, f64) {
        let n = self.data.len();
        if n > 1 {
            let mut sum_corr = 0.0;
            for i in 0..n {
                for j in 0..n {
                    let corr = self.pearson(&self.data[i], &self.data[j]);
                    sum_corr += corr.abs();
                }
            }
            self.c = sum_corr / (n * n) as f64;
        } else {
            self.c = 0.5;
        }
        self.f = 1.0 - self.c;
        let cons = (self.c + self.f - 1.0).abs() < 1e-10;
        let topo = if self.c > 0.8 { "toroidal".to_string() } else { "other".to_string() };
        let scaling = if self.c > 0.7 { "self-similar".to_string() } else { "linear".to_string() };
        (self.c, self.f, cons, topo, scaling, self.f * 0.5)
    }
}

fn main() {
    let data = vec![
        vec![1.0,2.0,3.0,4.0],
        vec![2.0,3.0,4.0,5.0],
        vec![5.0,6.0,7.0,8.0],
    ];
    let mut ucd = UCD::new(data);
    let (c, f, cons, topo, scaling, opt) = ucd.analyze();
    println!("C: {:.4}, F: {:.4}, conservation: {}, topology: {}, scaling: {}, opt: {:.4}",
             c, f, cons, topo, scaling, opt);
}
